src/common/src/array/data_chunk.rs

```diff
@@ -159,38 +169,41 @@ impl DataChunk {
 
     /// `cardinality` returns the number of visible tuples
     pub fn cardinality(&self) -> usize {
-        self.cardinality
+        match &self.vis2 {
+            Vis::Bitmap(b) => b.num_high_bits(),
+            Vis::Compact(len) => *len,
+        }
     }
 
     /// `capacity` returns physical length of any chunk column
     pub fn capacity(&self) -> usize {
-        if let Some(bitmap) = &self.visibility {
-            bitmap.len()
-        } else {
-            self.cardinality
+        match &self.vis2 {
+            Vis::Bitmap(b) => b.num_bits(),
+            Vis::Compact(len) => *len,
         }
     }
 
-    pub fn visibility(&self) -> &Option<Bitmap> {
-        &self.visibility
-    }
-
     #[must_use]
     pub fn with_visibility(&self, visibility: Bitmap) -> Self {
         DataChunk::new(self.columns.clone(), Some(visibility))
     }
 
+    pub fn visibility(&self) -> Option<&Bitmap> {
+        self.get_visibility_ref()
```

Comment:
Signature changed from `&Option<Bitmap>` to `Option<&Bitmap>`, and became redundant with `get_visibility_ref`. Not sure which name is better, but we should just keep one.

As for actual usage, the impact is minimal - only a few `.clone()` are changed to `.cloned()`.